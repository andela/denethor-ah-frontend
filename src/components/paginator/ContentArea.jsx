import React from 'react';
import uuid from 'uuid/v1';
import PropTypes from 'prop-types';
import FeedBottomSmallCard from '../feed/FeedBottomSmallCard';

export const ContentArea = ({ items }) => (
  <div className="paginator__content-area">
    {items.map(item => <FeedBottomSmallCard key={uuid()} {...item} /> )}
  </div>
);

ContentArea.propTypes = {
  items: PropTypes.array
}

export default ContentArea;
