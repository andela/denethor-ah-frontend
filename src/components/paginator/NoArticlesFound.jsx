import React from 'react';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimesCircle } from '@fortawesome/free-regular-svg-icons';

library.add(faTimesCircle);

const NoArticlesFound = () => (
  <div className='flex search-error'>
    <FontAwesomeIcon
      icon={faTimesCircle}
      size='3x'
      color='#ff0000'
    />
    &nbsp;&nbsp;&nbsp;
    <h2>No Article was found in this category</h2>
  </div>
);

export default NoArticlesFound;
