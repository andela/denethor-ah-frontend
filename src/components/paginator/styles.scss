.paginator {
  .paginator__content-area {
    display: flex;
    flex-flow: row wrap;
    justify-content: left;
    margin: 0 auto;
    width: 70%;
    .feed-bottom-container-small {
      width: 50%;
    }
  }
  .pagination {
    display: flex;
    list-style-type: none;
    justify-content: center;
    margin: 4rem 2rem;

    li {
      border: 1px solid #444;
      border-left: none;
      cursor: pointer;
      width: 2.5rem;
      &:first-child {
        border-left: 1px solid #444;
        border-top-left-radius: 3px;
        border-bottom-left-radius: 3px;
      }

      &:last-child {
        border-top-right-radius: 3px;
        border-bottom-right-radius: 3px;
      }

      &.previous, &.next {
        width: 5rem;
      }
      
      &.active {
        background: #444;
        color: white;
      }

      a {
        display: block;
        text-align: center;
        padding: 0 0.5rem;
        width: 100%;
      }
    }
  }
}

.search-error {
  height: 80px;
  width: 90%;
  justify-content: center;
  margin-top: 20%;
  margin-bottom: 15%;
 }

@media (max-width: 1080px) {
  .paginator {
    .paginator__content-area {
      width: 100%;
    }
  }
}

@media (max-width: 480px) {
  .paginator {
    .paginator__content-area {
      display: block;
      justify-content: center;
      width: 100%;
      .feed-bottom-container-small {
        width: 100%;
      }
    }
  }
}

.paginator-spinner-container {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100vh;
}
