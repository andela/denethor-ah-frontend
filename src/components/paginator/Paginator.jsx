import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PageNavigator from 'react-paginate';
import ContentArea from './ContentArea';
import './styles.scss';

export class Paginator extends Component {
  state = {
    offset: 0,
    limit: 10
  };

  componentDidMount () {
    const { props: { items: { length } }, state: { limit } } = this;
    this.setState({ pageCount: Math.ceil(length/limit) });
    window.scrollTo(0, 0);
  }

  handlePageChange = ({ selected: page }) => {
    const { limit } = this.state;
    this.setState({ offset: page * limit });
    window.scrollTo(0, 0);
  };

  render() {
    const { state: { offset, limit, pageCount }, props: { items }, handlePageChange } = this;

    const pageContent = items.filter((item, i) => (i >= offset) && (i < offset + limit));

    return (
      <div className="paginator">
        <ContentArea items={pageContent} />
        {
          pageCount > 1
            &&
          <PageNavigator
            previousLabel={'prev'}
            nextLabel={'next'}
            breakLabel={'...'}
            breakClassName={'break-me'}
            pageCount={pageCount}
            marginPagesDisplayed={2}
            pageRangeDisplayed={5}
            onPageChange={handlePageChange}
            containerClassName={'pagination'}
            subContainerClassName={'pages pagination'}
            activeClassName={'active'}
          />
        }
      </div>
    )
  }
}

Paginator.propTypes = {
  items: PropTypes.array
}

export default Paginator;
