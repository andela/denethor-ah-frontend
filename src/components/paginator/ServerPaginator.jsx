import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PageNavigator from 'react-paginate';
import ContentArea from './ContentArea';
import axios from '../../utils/axiosConfig';
import { extractImageFromBody } from '../../utils/imageExtractor';
import { toTimeFromNow } from '../../utils/dateTime';
import NoArticlesFound from './NoArticlesFound';
import Spinner from '../spinner/Spinner';
import './styles.scss';

export class Paginator extends Component {
  state = {};

  api = process.env.API_ROOT_URL;

  async componentDidMount () {
    const { url } = this.props;
    this.setState({ loading: true });
    window.scrollTo(0, 0);
    const { data: { data: items, count } } = await axios.get(`${this.api}${url}`);
    this.setState({ loading: false });
    this.setState({ url, items, pageCount: Math.ceil(count/10) });
  }

  async componentDidUpdate ({ url: prevUrl }) {
    const { url } = this.props;
    if (url !== prevUrl) {
      this.setState({ loading: true });
      window.scrollTo(0, 0);
      const { data: { data: items, count } } = await axios.get(`${this.api}${url}`);
      this.setState({ loading: false });
      this.setState({ url, items, pageCount: Math.ceil(count/10) });
    }
  }

  handlePageChange = async ({ selected: page }) => {
    const { url } = this.state;
    this.setState({ loading: true });
    window.scrollTo(0, 0);
    const { data: { data: items } } = await axios.get(`${this.api}${url}&n=${page}`);
    this.setState({ loading: false });
    this.setState({ items });
  };

  render() {
    const { state: { loading, pageCount, items = [] }, handlePageChange } = this;

    const articles = items.map(article => ({
      ...article,
      featuredImage: extractImageFromBody(article.body),
      dateCreated: toTimeFromNow(article.createdAt)
    }));

    return loading
      ? <div className='paginator-spinner-container'>
          <Spinner loading={loading} />
        </div>
      : articles[0]
        ? (
          <div className="paginator">
            <ContentArea items={articles} />
            {
              pageCount > 1
                &&
                <PageNavigator
                  previousLabel={'prev'}
                  nextLabel={'next'}
                  breakLabel={'...'}
                  breakClassName={'break-me'}
                  pageCount={pageCount}
                  marginPagesDisplayed={2}
                  pageRangeDisplayed={5}
                  onPageChange={handlePageChange}
                  containerClassName={'pagination'}
                  subContainerClassName={'pages pagination'}
                  activeClassName={'active'}
                />
            }
          </div>
      ) : (
        <NoArticlesFound />
      )
  }
}

Paginator.propTypes = {
  url: PropTypes.string
}

export default Paginator;
