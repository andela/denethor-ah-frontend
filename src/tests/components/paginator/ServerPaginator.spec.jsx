import React from 'react';
import { shallow } from 'enzyme';
import { Paginator } from '../../../components/paginator/ServerPaginator';
import axios from '../../../utils/axiosConfig';

jest.mock('../../../utils/axiosConfig.js');

test('Should render static component', () => {
  axios.get.mockResolvedValue({ data: { data: { items: {}, count: 4 } } });
  const wrapper = shallow(<Paginator url='string' />);
  expect(wrapper).toMatchSnapshot();
});
