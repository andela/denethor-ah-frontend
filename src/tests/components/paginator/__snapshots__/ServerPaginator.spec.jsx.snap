// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<div
  className="paginator-spinner-container"
>
  <Spinner
    loading={true}
  />
</div>
`;
